runtime/lua/vim/_helptags.lua

```diff
@@ -0,0 +1,198 @@
+local M = {}
+
+local ts = vim.treesitter
+local async = vim._async
+local query = ts.query.parse('vimdoc', '(tag (word) @tagname)')
+
+--- @alias Tag { [1]: string, [2]: string, [3]: string}[] tuple of tag, file, and search command
+
+---Read file in libuv callback-style
+---@param path string
+---@param callback fun(string?, string?): string?, string?
+local function read_file_uv(path, callback)
+  vim.uv.fs_open(path, 'r', 438, function(err_open, fd)
+    if err_open then
+      return callback(nil, err_open)
+    end
+
+    vim.uv.fs_fstat(fd, function(err_stat, stat)
+      if err_stat then
+        vim.uv.fs_close(fd)
+        return callback(nil, err_stat)
+      end
+
+      vim.uv.fs_read(fd, stat.size, 0, function(err_read, data)
+        vim.uv.fs_close(fd)
+        if err_read then
+          return callback(nil, err_read)
+        end
+        callback(data, nil)
+      end)
+    end)
+  end)
+end
+
+--- Extract all tag definitions from a help file.
+--- @async
+--- @param filename string Help file with tags.
+--- @return Tag[] # List of tags. Empty if file is not readable.
+local function extract_file_tags(filename)
+  local tags = {}
+  local source, err = async.await(2, read_file_uv, filename)
+  if err or source == nil then
+    async.await(vim.schedule)
+    vim.notify(('E153: Unable to open %s for reading'):format(filename), vim.log.levels.ERROR)
+    return tags
+  end
```

Comment:
this is what the original code did. I can also just `assert` that the file is read if that's preferred.